### PR TITLE
feat(log): Show the current network upgrade in progress logs

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

It takes almost 6 hours for Zebra to sync all its checkpoints, so we need to be able to split CI jobs by network upgrade as well as checkpoint/full validation.

This is part of ticket #4661, it also includes a refactor I did as part of loading blocks from a file.

## Solution

Features:
- Show the network upgrade in progress logs
- Log when Zebra verifies the final checkpoint

Fixes:
- Log elapsed time using `humantime` truncated to whole seconds, rather than raw seconds and microseconds
- Fix incorrect negative sign on elapsed time
- Remove the space before the percentage sign when showing percent progress

Refactors:
- Move the progress task into its own module

## Review

This is a high priority because I need it for ticket #4661.

### Reviewer Checklist

  - [ ] Progress logs contain network upgrade
  - [ ] Time is formatted nicely
  - [ ] Existing tests pass

### CODE REVIEW

File: zebrad/src/components/sync/progress.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+        if let Some(estimated_height) =
+            latest_chain_tip.estimate_network_chain_tip_height(network, now)
+        {
+            // The estimate/actual race doesn't matter here,
+            // because we're only using it for metrics and logging.
+            let current_height = latest_chain_tip
+                .best_tip_height()
+                .expect("unexpected empty state: estimate requires a block height");
+            let network_upgrade = NetworkUpgrade::current(network, current_height);
```
Comment (teor2345) [inline zebrad/src/components/sync/progress.rs:116]: Most of this code was moved, but this is new in this PR.

File: zebrad/src/components/sync/progress.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+            // Work out the sync progress towards the estimated tip.
+            let sync_progress = f64::from(current_height.0) / f64::from(estimated_height.0);
+            let sync_percent = format!(
+                "{:.frac$}%",
```
Comment (teor2345) [inline zebrad/src/components/sync/progress.rs:121]: I deleted a space between `{:.frac$}` and `%`.

File: zebrad/src/components/sync/progress.rs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+            // Work out how long it has been since the state height has increased.
+            //
+            // Non-finalized forks can decrease the height, we only want to track increases.
+            if current_height > last_state_change_height {
+                last_state_change_height = current_height;
+                last_state_change_time = now;
+            }
+
+            let time_since_last_state_block_chrono =
+                now.signed_duration_since(last_state_change_time);
```
Comment (teor2345) [inline zebrad/src/components/sync/progress.rs:137]: I fixed this calculation, it used to return negative times.

File: zebrad/src/components/sync/progress.rs [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+            );
+
+            if time_since_last_state_block_chrono > min_state_block_interval {
+                // The state tip height hasn't increased for a long time.
+                //
+                // Block verification can fail if the local node's clock is wrong.
+                warn!(
+                    %sync_percent,
+                    ?current_height,
+                    ?network_upgrade,
```
Comment (teor2345) [inline zebrad/src/components/sync/progress.rs:151]: I added the network upgrade to every log where we have a current height.

File: zebra-consensus/src/checkpoint.rs [hunk 5]
```diff
@@ -464,6 +464,11 @@ where
         if verified_height == self.checkpoint_list.max_height() {
             metrics::gauge!("checkpoint.verified.height", verified_height.0 as f64);
             self.verifier_progress = FinalCheckpoint;
+
+            tracing::info!(
+                final_checkpoint_height = ?verified_height,
+                "verified final checkpoint: starting full validation",
+            );
```
Comment (teor2345) [inline zebra-consensus/src/checkpoint.rs:471]: I needed this to split checkpoint and full validation logs into different CI jobs

**Review by upbqdn — APPROVED** [review]
All looks great.

By reviewing this PR, I learned to compare only parts of different files against each other so that I can see what's changed, and what's just moved.

**Comment by teor2345** [comment]
> All looks great.
> 
> By reviewing this PR, I learned to compare only parts of different files against each other so that I can see what's changed, and what's just moved.

Yeah I find `git diff --color-moved` really helpful, I wish GitHub would do it automatically.

**Pull request merged**